### PR TITLE
#8 - Create Controllers

### PR DESCRIPTION
**Pull request by delanohelio**

### CODE REVIEW

File: src/main/java/com/nanuvem/metagui/server/api/Context.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+	@Id
+	@GeneratedValue
+	private Long id;
+	@Column(nullable = false, unique = true)
+	private String name;
+	@NotNull
+	private WidgetType type;
+	@JsonIgnore
+	@ManyToMany(mappedBy = "requiredContexts")
```
Comment (rodrigovilar) [inline src/main/java/com/nanuvem/metagui/server/api/Context.java]: Use @ManyToOne

File: src/main/java/com/nanuvem/metagui/server/api/Widget.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+import javax.persistence.OneToMany;
+import javax.persistence.Table;
+import javax.persistence.UniqueConstraint;
+import javax.validation.constraints.NotNull;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+
+
+@Entity
+@Table(uniqueConstraints=@UniqueConstraint(columnNames={"name", "version"}))
```
Comment (rodrigovilar) [inline src/main/java/com/nanuvem/metagui/server/api/Widget.java:16]: Verify if this unique constraint works.
It did not work for me some years ago.

File: src/main/java/com/nanuvem/metagui/server/api/Widget.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+	private Long id;
+	@NotNull
+	private String name;
+	@NotNull
+	private Long version;
+	private String code;
+	@NotNull
+	private WidgetType type;
+	private String configuration;
+	@ManyToMany(fetch = FetchType.EAGER)
```
Comment (rodrigovilar) [inline src/main/java/com/nanuvem/metagui/server/api/Widget.java]: Change to @OneToMany

File: src/main/java/com/nanuvem/metagui/server/api/Widget.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+	private List<Context> requiredContexts;
+	@JsonIgnore
+	@OneToMany(mappedBy = "widget")
```
Comment (rodrigovilar) [inline src/main/java/com/nanuvem/metagui/server/api/Widget.java]: Remove this rules field.

File: src/main/java/com/nanuvem/metagui/server/context/ContextContainer.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+import java.util.List;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import com.nanuvem.metagui.server.api.Context;
+
+@Component
+public class ContextContainer {
```
Comment (rodrigovilar) [inline src/main/java/com/nanuvem/metagui/server/context/ContextContainer.java]: Rename to ContextService

File: src/main/java/com/nanuvem/metagui/server/rules/RulesContainer.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+import org.springframework.stereotype.Component;
+
+import com.nanuvem.metagui.server.api.Context;
+import com.nanuvem.metagui.server.api.Rule;
+import com.nanuvem.metagui.server.api.Widget;
+import com.nanuvem.metagui.server.context.ContextContainer;
+import com.nanuvem.metagui.server.widgets.WidgetContainer;
+
+@Component
+public class RulesContainer {
```
Comment (rodrigovilar) [inline src/main/java/com/nanuvem/metagui/server/rules/RulesContainer.java]: Rename to RuleService

File: src/main/java/com/nanuvem/metagui/server/rules/RulesContainer.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+	public List<Rule> getAllRules() {
+		return ruleRepository.findAll();
+	}
+	
+	public List<Rule> getAllRulesByVersionGreaterThan(Long version) {
+		return ruleRepository.findByVersionGreaterThan(version);
+	}
+	
+	public Rule saveRule(@Valid Rule rule) {
+		rule.setVersion(++version);
```
Comment (rodrigovilar) [inline src/main/java/com/nanuvem/metagui/server/rules/RulesContainer.java]: You need to update version, every time this component is instantiated.
Maybe this solution is not thread safe.

File: src/main/java/com/nanuvem/metagui/server/widgets/WidgetContainer.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+	
+	public Widget saveWidget(Widget widget) {
+		widget.setVersion(1l);
+		Widget widgetSameName = getWidgetByName(widget.getName());
+		if(widgetSameName != null) {
+			widget.setVersion(widgetSameName.getVersion() + 1);
+		}
+		if(widget.getRequiredContexts() != null) {
+			for(Context context : widget.getRequiredContexts()) {
+				context.setId(contextContainer.getOrCreateContext(context).getId());
```
Comment (rodrigovilar) [inline src/main/java/com/nanuvem/metagui/server/widgets/WidgetContainer.java]: Replace getOrCreateContext by createContext

File: src/main/java/com/nanuvem/metagui/server/widgets/WidgetContainer.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+import org.springframework.stereotype.Component;
+import org.springframework.stereotype.Service;
+
+import com.nanuvem.metagui.server.api.Context;
+import com.nanuvem.metagui.server.api.Widget;
+import com.nanuvem.metagui.server.context.ContextContainer;
+
+@Service
+@Component
+public class WidgetContainer {
```
Comment (rodrigovilar) [inline src/main/java/com/nanuvem/metagui/server/widgets/WidgetContainer.java]: Rename to WidgetService

File: src/main/java/com/nanuvem/metagui/server/widgets/controller/WidgetController.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+	
+	@RequestMapping(method = RequestMethod.GET)
+	@ResponseBody
+	public ResponseEntity<List<Widget>> getAll() {
+		return new ResponseEntity<List<Widget>>(widgetContainer.getAll(), HttpStatus.OK);
+	}
+	
+	@RequestMapping(method = RequestMethod.POST)
+	@ResponseBody
+	public ResponseEntity<Widget> createEntityWidget(@RequestBody String input) {
```
Comment (rodrigovilar) [inline src/main/java/com/nanuvem/metagui/server/widgets/controller/WidgetController.java]: This works only for EntityWidget?

File: src/test/java/com/nanuvem/metagui/server/gui/WidgetsAndRulesControllerTest.java [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+	
+	@DirtiesContext
+	@Test
+	public void testCreateWidgets() throws Exception {
+		Widget widget = TestCreatorHelper.createSimpleWidget("fooName", "fooCode", WidgetType.Property);
+		
+		Map<String, Object> widgetInstanceMap = objectToMap(widget);
+		widgetInstanceMap.remove("id");
+		widgetInstanceMap.put("version", 1);
+		widgetInstanceMap.put("type", widget.getType().name());
```
Comment (rodrigovilar) [inline src/test/java/com/nanuvem/metagui/server/gui/WidgetsAndRulesControllerTest.java]: create a helper method for lines 73-75

**Comment by rodrigovilar** [comment]
Review closed.

File: src/main/java/com/nanuvem/metagui/server/api/Widget.java (hunk 2 above)
Comment (delanohelio) [inline src/main/java/com/nanuvem/metagui/server/api/Widget.java:16]: It's worked for me!

Comment (rodrigovilar) [inline src/main/java/com/nanuvem/metagui/server/api/Widget.java:16]: Ok!

File: src/main/java/com/nanuvem/metagui/server/rules/RuleService.java [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+	@Autowired
+	private WidgetService widgetContainer;
+	
+	private Long version;
+	
+	private void updateVersion() {
+		version = 1l;
+		List<Rule> rules = ruleRepository.findAll(new Sort(Sort.Direction.DESC, "version"));
+		if(rules.size() > 0) {
+			version = rules.get(0).getVersion();
```
Comment (rodrigovilar) [inline src/main/java/com/nanuvem/metagui/server/rules/RuleService.java:34]: In another task, do a select max(version) query

**Pull request merged**